neqo-transport/src/connection/mod.rs

```diff
@@ -1828,13 +1825,24 @@ impl Connection {
                     let read = self.crypto.streams.read_to_end(space, &mut buf);
                     qdebug!("Read {} bytes", read);
                     self.handshake(now, space, Some(&buf))?;
+                    let mut tokens = Vec::new();
+                    if let Agent::Client(ref mut c) = self.crypto.tls {
+                        while let Some(t) = c.resumption_token() {
+                            tokens.push(t);
+                        }
+                    }
+                    while let Some(t) = tokens.pop() {
+                        self.last_token = Some(t);
+                        self.create_resumption_token();
+                    }
                 } else {
                     // If we get a useless CRYPTO frame send outstanding CRYPTO frames again.
                     self.crypto.resend_unacked(space);
                 }
             }
             Frame::NewToken { token } => {
                 self.new_token.save_token(token);
+                self.create_resumption_token();
```

Comment:
Ahh, I see how you are getting 4 token events.  I don't think that we want this.  See the meta-comment.